Corona-Warn-App/src/main/java/de/rki/coronawarnapp/vaccination/core/repository/ValueSetsRepository.kt

```diff
@@ -19,55 +26,78 @@ import javax.inject.Inject
 class ValueSetsRepository @Inject constructor(
     private val vaccinationServer: VaccinationServer,
     private val valueSetsStorage: ValueSetsStorage,
-    @AppScope private val scope: CoroutineScope
+    @AppScope private val scope: CoroutineScope,
+    dispatcherProvider: DispatcherProvider
 ) {
 
-    val latestValueSet: Flow<VaccinationValueSet> = valueSetsStorage.valueSet.flow.distinctUntilChanged()
+    private fun Flow<VaccinationValueSet>.distinctUntilChangedByHash() = distinctUntilChangedBy { it.hashCode() }
 
-    fun triggerUpdateValueSet(languageCode: Locale) = scope.launch {
+    private val internalData: HotDataFlow<VaccinationValueSet> = HotDataFlow(
+        loggingTag = TAG,
+        scope = scope,
+        coroutineContext = dispatcherProvider.IO,
+        sharingBehavior = SharingStarted.Lazily,
+        startValueProvider = {
+            (valueSetsStorage.vaccinationValueSet ?: createEmptyValueSet())
+                .also { Timber.v("Loaded initial value set %s", it) }
+        }
+    )
+
+    init {
+        internalData.data
+            .distinctUntilChangedByHash()
+            .onStart { Timber.d("Observing value set") }
+            .onEach { valueSetsStorage.vaccinationValueSet = it }
+            .catch { Timber.e(it, "Storing new value set failed.") }
+            .launchIn(scope + dispatcherProvider.IO)
+    }
+
+    val latestValueSet: Flow<VaccinationValueSet> = internalData.data.distinctUntilChangedByHash()
+
+    fun triggerUpdateValueSet(languageCode: Locale) {
         Timber.d("triggerUpdateValueSet(languageCode=%s)", languageCode)
+
+        internalData.updateAsync(
+            onUpdate = { updateValueSet(languageCode = languageCode) },
+            onError = { Timber.e(it, "Updating value set failed") }
+        )
+    }
+
+    private suspend fun updateValueSet(languageCode: Locale): VaccinationValueSet {
+        Timber.v("updateValueSet(languageCode=%s)", languageCode)
         var valueSet = vaccinationServer.getVaccinationValueSets(languageCode = languageCode)
+            ?: valueSetsStorage.vaccinationValueSet
```

Comment:
Try to stay within the threadsafe execution of `HotDataFlow`, the `valueSetsStorage.vaccinationValueSet` accesses the current data from storage again, the fallback here should be the original data within your execution scope though, i.e.:
`updateAsync.onUpdate { oldData -> newData }`, if there is an error just do ` {oldData -> oldData}`

Reply:
"[...]_updateBlocking/update, alle diese update operationen auf dem HotDataFlow Datensatz sind threadsafe_"
Did I get it wrong?